src/sdk/python/rtdip_sdk/pipelines/monitoring/spark/data_quality/identify_missing_data_pattern.py

```diff
@@ -0,0 +1,310 @@
+        min_time, max_time = df_sorted.agg(
+            F.min("EventTime"), F.max("EventTime")
+        ).first()
+        if not min_time or not max_time:
+            self.logger.info("Generated 0 expected times based on patterns.")
+            self.logger.info("DataFrame is empty. No missing patterns to detect.")
+            return self.df
+        # Generate all expected times based on patterns and frequency
+        expected_times_df = self._generate_expected_times(min_time, max_time)
+        # Identify missing patterns by left joining expected times with actual EventTimes within tolerance
+        missing_patterns_df = self._find_missing_patterns(expected_times_df, df_sorted)
+        self._log_missing_patterns(missing_patterns_df)
+        return self.df
+
+    def _validate_inputs(self):
+        if self.frequency not in ["minutely", "hourly"]:
+            error_msg = "Frequency must be either 'minutely' or 'hourly'."
+            self.logger.error(error_msg)
+            raise ValueError(error_msg)
+        for pattern in self.patterns:
+            if self.frequency == "minutely":
+                if "second" not in pattern:
+                    raise ValueError(
+                        "Each pattern must have a 'second' key for 'minutely' frequency."
+                    )
+                if pattern.get("second", 0) >= 60:
+                    raise ValueError(
+                        "For 'minutely' frequency, 'second' must be less than 60."
+                    )
+                if "minute" in pattern or "hour" in pattern:
+                    raise ValueError(
+                        "For 'minutely' frequency, pattern should not contain 'minute' or 'hour'."
+                    )
+            elif self.frequency == "hourly":
+                if "minute" not in pattern or "second" not in pattern:
+                    raise ValueError(
+                        "Each pattern must have 'minute' and 'second' keys for 'hourly' frequency."
+                    )
+                if pattern.get("minute", 0) >= 60:
+                    raise ValueError(
+                        "For 'hourly' frequency, 'minute' must be less than 60."
+                    )
+                if "hour" in pattern:
+                    raise ValueError(
+                        "For 'hourly' frequency, pattern should not contain 'hour'."
+                    )
+        try:
+            self.tolerance_ms = parse_time_string_to_ms(self.tolerance)
+            self.tolerance_seconds = self.tolerance_ms / 1000
+            self.logger.info(
+                f"Using tolerance: {self.tolerance_ms} ms ({self.tolerance_seconds} seconds)"
+            )
+        except ValueError as e:
+            error_msg = f"Invalid tolerance format: {self.tolerance}"
+            self.logger.error(error_msg)
+            raise ValueError(error_msg) from e
+
+    def _generate_expected_times(self, min_time, max_time) -> PySparkDataFrame:
```

Comment:
long function - let's divide into smaller sub functions to be called from this

Reply:
handled in 90bec2d